components/brave_federated/communication_adapter.cc

```diff
@@ -0,0 +1,193 @@
+            "Reports are NOT generated when the user is on battery power "
+            "or when the device is on a metered network (not WiFi/Ethernet). "
+          data:
+            "Simple federated model updates based off toy task and locally "
+            "generated synthetic data."
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          setting:
+            "This experimental feature is only enabled in Nightly builds and"
+            "can be disabled via the 'brave://flags/#brave-federated' "
+          policy_exception_justification:
+            "Not implemented."
+        }
+    )");
+}
+
+}  // namespace
+
+CommunicationAdapter::CommunicationAdapter(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    const net::BackoffEntry::Policy reconnect_policy,
+    const net::BackoffEntry::Policy request_task_policy)
+    : url_loader_factory_(std::move(url_loader_factory)) {
+  reconnect_policy_ =
+      std::make_unique<const net::BackoffEntry::Policy>(reconnect_policy);
+  reconnect_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(reconnect_policy_.get());
+
+  request_task_policy_ =
+      std::make_unique<const net::BackoffEntry::Policy>(request_task_policy);
+  request_task_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(request_task_policy_.get());
+}
+
+CommunicationAdapter::~CommunicationAdapter() = default;
+
+void CommunicationAdapter::GetTasks(GetTaskCallback callback) {
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->url = GURL(features::GetFederatedLearningTaskEndpoint());
+  request->headers.SetHeader("Content-Type", "application/protobuf");
+  request->headers.SetHeader("Accept", "application/protobuf");
+  request->headers.SetHeader("X-Brave-FL-Federated-Learning", "?1");
+  request->credentials_mode = network::mojom::CredentialsMode::kOmit;
+  request->load_flags = net::LOAD_DO_NOT_SAVE_COOKIES;
+  request->method = net::HttpRequestHeaders::kPostMethod;
+
+  VLOG(2) << "Requesting tasks list " << request->method << " " << request->url;
+
+  const std::string& payload = BuildGetTasksPayload();
+
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(request), GetNetworkTrafficAnnotationTag());
+  url_loader_->AttachStringForUpload(payload, "application/protobuf");
+  url_loader_->DownloadToString(
+      url_loader_factory_.get(),
+      base::BindOnce(&CommunicationAdapter::OnGetTasks,
+                     weak_factory_.GetWeakPtr(), std::move(callback)),
+      kMaxFederatedServerResponseSizeBytes);
+}
+
+void CommunicationAdapter::OnGetTasks(
+    GetTaskCallback callback,
+    const std::unique_ptr<std::string> response_body) {
+  bool failed_request =
+      !url_loader_->ResponseInfo() || !url_loader_->ResponseInfo()->headers;
+  reconnect_backoff_entry_->InformOfRequest(!failed_request);
+  int64_t reconnect_delay_in_seconds =
```

Comment:
ideally we should not serialize/deserialize base::TimeDelta, we should just use base::TimeDelta